src/fsharp/infos.fs

```diff
@@ -1316,10 +1328,27 @@ type MethInfo =
                     | Some b -> ReflectedArgInfo.Quote b
                     | None -> ReflectedArgInfo.None
                 let isOutArg = HasFSharpAttribute g g.attrib_OutAttribute argInfo.Attribs && isByrefTy g ty
-                let isOptArg = HasFSharpAttribute g g.attrib_OptionalArgumentAttribute argInfo.Attribs
-                // Note: can't specify caller-side default arguments in F#, by design (default is specified on the callee-side) 
-                let optArgInfo = if isOptArg then CalleeSide else NotOptional
-                
+                let isCalleeSideOptArg = HasFSharpAttribute g g.attrib_OptionalArgumentAttribute argInfo.Attribs
+                let isCallerSideOptArg = HasFSharpAttributeOpt g g.attrib_OptionalAttribute argInfo.Attribs
+                let optArgInfo = 
+                    if isCalleeSideOptArg then 
+                        CalleeSide 
+                    elif isCallerSideOptArg then
+                        let defaultParameterValueAttribute = TryFindFSharpAttributeOpt g g.attrib_DefaultParameterValueAttribute argInfo.Attribs
+                        match defaultParameterValueAttribute with
+                        | None -> 
+                            // Do a type-directed analysis of the type to determine the default value to pass.
+                            // Similar rules as OptionalArgInfo.FromILParameter are applied here, except for the COM and byref-related stuff.
+                            CallerSide (if isObjTy g ty then MissingValue else DefaultValue)
+                        | Some attr -> 
+                            let defaultValue = OptionalArgInfo.ValueOfDefaultParameterValueAttrib attr
+                            match defaultValue with
+                            | Some (Expr.Const (_, m, typ)) when not (typeEquiv g typ ty) -> 
+                                error(Error(FSComp.SR.DefaultParameterValueNotAppropriateForArgument(), m)); NotOptional
```

Comment:
The problem is that then the compiler can generate unverifiable IL - the default value is put on the stack and the method is called, but the value on the stack will have the type of the default value, which does not correspond to the method argument's type.

So our options are:
- Keep it as an error
- Make it a warning, but ignore the default value (i.e. infer the default value from the type)
- Make it a warning, ignore the default value and the optional argument (i.e. won't be able to omit this argument) - because inferring the default value from the type can have some surprising consequences.